tket2/src/rewrite_space/factory/kak.rs

```diff
@@ -0,0 +1,255 @@
+            for expanded_walker in searching_walker.walker.expand(&wire_to_search, Direction::Outgoing) {
+                
+                // Get the new expanded version of the wire to search along.
+                let expanded_wire = expanded_walker.get_wire(
+                    frontier_node, 
+                    IncomingPort::from(0),
+                );
+
+                // Get the node at the end of this newly expanded wire.
+                let (expanded_node, _) = expanded_walker.wire_pinned_outport(&expanded_wire).expect("error");
+
+                // Get the type of operation this node implements
+                let ext_op = expanded_walker.as_hugr_view().get_optype(expanded_node).as_extension_op().expect("Not extension op");
+                let tket2_op = Tk2Op::from_extension_op(ext_op).expect("Not TKET2 OP");
+
+                // If it it not a quantum operation then there is no need to continue
+                // search as this is not a 2q sub circuit.
+                if !tket2_op.is_quantum() {
+                    continue;
+                
+                // If it is a one qubit operation then we should continue searching.
+                // We add the searching walker back to the queue, updated as required.
+                } else if n_qubits(expanded_node, expanded_walker.as_hugr_view()) == 1{
+                    
+                    searching_walker.pattern.wires.push(expanded_wire);
+                    searching_walker.frontier_wires.push(expanded_walker.get_wire(expanded_node, OutgoingPort::from(0)));
+                    searching_walker.walker = expanded_walker;
+                    
+                    searching_walker_list.push(searching_walker);
+                
+                // If it is a 2q operation then the action depends on which
+                // qubits the gate acts on.
+                } else if walker.as_hugr_view().get_optype(expanded_node) == Tk2Op::CZ.into() {
+
+                    // If a 2q gate has already been found by this searcher,
+                    // then we need to check if we have found the same one again.
+                    if let Some(two_qubit_node) = searching_walker.two_qubit_node {
+
+                        // If the one just found is the same 2q gate that this searcher has already
+                        // found, then we have reached the end of a 2q subcircuit.
+                        // We add the pattern to our matches, and add the
+                        // search back to the queue to continue exploring
+                        // along the outputs of the new 2q gate.
+                        // If it is not the same then we don't need to do
+                        // anything as this is not a 2q sub circuit.
+                        if two_qubit_node == expanded_node {
+
+                            searching_walker.pattern.wires.push(expanded_wire);
+
+                            matches.push((searching_walker.pattern, searching_walker.walker));
```

Comment:
I suspect the compiler must be complaining here about using `searching_walker.pattern` and `searching_walker.walker` in more than one place (more specifically, it will say "use of moved value" further below). You must clone the data:

```suggestion
                            matches.push((searching_walker.pattern.clone(), searching_walker.walker.clone()));
```

That being said, this will add all possible subcircuits (not just the maximal ones). Is this what you want? Might indeed be the right way to go provided we don't get far too many matches.

Reply:
Thanks!
Adding all matches was indeed what I was going for as a first try.